File: lwk_bindings/tests/bindings/multisig.py
from lwk import *

node = LwkTestEnv() # launch electrs and elementsd

# ANCHOR: multisig-setup
network = Network.testnet()
network = Network.regtest_default() # ANCHOR: ignore
# Derivation for multisig
bip = Bip.new_bip87()

# Alice creates their signer and gets the xpub
signer_a = Signer.random(network)
xpub_a = signer_a.keyorigin_xpub(bip);

# Bob creates their signer and gets the xpub
signer_b = Signer.random(network)
xpub_b = signer_b.keyorigin_xpub(bip);

# Carol, who acts as a coordinator, creates their signer and gets the xpub
signer_c = Signer.random(network)
xpub_c = signer_c.keyorigin_xpub(bip);

# Carol generates a random SLIP77 descriptor blinding key
import os
slip77_rand_key = os.urandom(32).hex()
desc_blinding_key = f"slip77({slip77_rand_key})"

# Carol uses the collected xpubs and the descriptor blinding key to create
# the 2of3 descriptor
threshold = 2;
desc = f"ct({desc_blinding_key},elwsh(multi({threshold},{xpub_a}/<0;1>/*,{xpub_b}/<0;1>/*,{xpub_c}/<0;1>/*)))"
# Validate the descriptor string
wd = WolletDescriptor(desc)
# ANCHOR_END: multisig-setup

# ANCHOR: multisig-receive
# Carol creates the wollet
wollet_c = Wollet(network, wd, datadir=None)

# With the wollet, Carol can obtain addresses, transactions and balance
addr = wollet_c.address(None);
txs = wollet_c.transactions();
balance = wollet_c.balance();

# Update the wollet state
url = "https://blockstream.info/liquidtestnet/api"
client = EsploraClient(url, network)
client = ElectrumClient.from_url(node.electrum_url())  # ANCHOR: ignore

update = client.full_scan(wollet_c)
wollet_c.apply_update(update)
# ANCHOR_END: multisig-receive

# Receive some funds
client = ElectrumClient.from_url(node.electrum_url())
txid = node.send_to_address(wollet_c.address(0).address(), 10_000, asset=None)
wollet_c.wait_for_tx(txid, client)

# ANCHOR: multisig-send
# Carol creates a transaction send few sats to a certain address
address = "<address>"
address = node.get_new_address() # ANCHOR: ignore
sats = 1000
lbtc = network.policy_asset()

b = network.tx_builder()
b.add_recipient(address, sats, lbtc)
pset = b.finish(wollet_c)

pset = signer_c.sign(pset)

# Carol sends the PSET to Bob
# Bob wants to analyze the PSET before signing, thus he creates a wollet
wd = WolletDescriptor(desc)
wollet_b = Wollet(network, wd, datadir=None)
update = client.full_scan(wollet_b)
wollet_b.apply_update(update)
# Then Bob uses the wollet to analyze the PSET
details = wollet_b.pset_details(pset)
# PSET has a reasonable fee
assert details.balance().fee() < 100
# PSET has a signature from Carol
fingerprints_has = details.fingerprints_has()
assert len(fingerprints_has) == 1
assert signer_c.fingerprint() in fingerprints_has
# PSET needs a signature from either Bob or Carol
fingerprints_missing = details.fingerprints_missing()
assert len(fingerprints_missing) == 2
assert signer_a.fingerprint() in fingerprints_missing
assert signer_b.fingerprint() in fingerprints_missing
# PSET has a single recipient, with data matching what was specified above
assert len(details.balance().recipients()) == 1
recipient = details.balance().recipients()[0]
assert str(recipient.address()) == str(address)
assert recipient.asset() == lbtc
assert recipient.value() == sats

# Bob is satisified with the PSET and signs it
pset = signer_b.sign(pset)

# Bob sends the PSET back to Carol
# Carol checks that the PSET has enough signatures
details = wollet_c.pset_details(pset)
fingerprints_has = details.fingerprints_has()
assert len(fingerprints_has) == 2

# Carol finalizes the PSET and broadcast the transaction
tx = pset.finalize()
txid = client.broadcast(tx)
# ANCHOR_END: multisig-send

File: lwk_bindings/go/multisig.go
package main

import (
	"fmt"
	"log"

	"lwk-go/lwk"
)

func main() {
	node := lwk.NewLwkTestEnv()

	// ANCHOR: multisig-setup
	network := lwk.NetworkTestnet()
	network = node.Network() // ANCHOR: ignore

	// Derivation for multisig
	bip := lwk.BipNewBip87()

	// Alice creates their signer and gets the xpub
	signerA, err := lwk.SignerRandom(network)
	if err != nil {
		log.Fatal(err)
	}
	xpubA, err := signerA.KeyoriginXpub(bip)
	if err != nil {
		log.Fatal(err)
	}

	// Bob creates their signer and gets the xpub
	signerB, err := lwk.SignerRandom(network)
	if err != nil {
		log.Fatal(err)
	}
	xpubB, err := signerB.KeyoriginXpub(bip)
	if err != nil {
		log.Fatal(err)
	}

	// Carol, who acts as a coordinator, creates their signer and gets the xpub
	signerC, err := lwk.SignerRandom(network)
	if err != nil {
		log.Fatal(err)
	}
	xpubC, err := signerC.KeyoriginXpub(bip)
	if err != nil {
		log.Fatal(err)
	}

	// Carol generates a random SLIP77 descriptor blinding key
	slip77RandKey := "<random-64-hex-chars>"
	slip77RandKey = "1111111111111111111111111111111111111111111111111111111111111111" // ANCHOR: ignore
	descBlindingKey := fmt.Sprintf("slip77(%s)", slip77RandKey)

	// Carol uses the collected xpubs and the descriptor blinding key to create
	// the 2of3 descriptor
	threshold := 2
	descStr := fmt.Sprintf("ct(%s,elwsh(multi(%d,%s/<0;1>/*,%s/<0;1>/*,%s/<0;1>/*)))", descBlindingKey, threshold, xpubA, xpubB, xpubC)

	// Validate the descriptor string
	wd, err := lwk.NewWolletDescriptor(descStr)
	if err != nil {
		log.Fatal(err)
	}
	// ANCHOR_END: multisig-setup

	// ANCHOR: multisig-receive
	// Carol creates the wollet
	wolletC, err := lwk.NewWollet(network, wd, nil)
	if err != nil {
		log.Fatal(err)
	}

	// With the wollet, Carol can obtain addresses, transactions and balance
	addr, err := wolletC.Address(nil)
	if err != nil {
		log.Fatal(err)
	}
	txs, err := wolletC.Transactions()
	if err != nil {
		log.Fatal(err)
	}
	balance, err := wolletC.Balance()
	if err != nil {
		log.Fatal(err)
	}

	_ = addr    // ANCHOR: ignore
	_ = txs     // ANCHOR: ignore
	_ = balance // ANCHOR: ignore

	// Update the wollet state
	url := "https://blockstream.info/liquidtestnet/api"
	client, err := lwk.NewEsploraClient(url, network)
	if err != nil {
		log.Fatal(err)
	}

	client, err = lwk.NewEsploraClient(node.EsploraUrl(), network) // ANCHOR: ignore
	if err != nil {                                                // ANCHOR: ignore
		log.Fatal(err) // ANCHOR: ignore
	} // ANCHOR: ignore

	update, err := client.FullScan(wolletC)
	if err != nil {
		log.Fatal(err)
	}
	if update != nil {
		if err := wolletC.ApplyUpdate(*update); err != nil {
			log.Fatal(err)
		}
	}
	// ANCHOR_END: multisig-receive

	// Receive some funds
	addr0, err := wolletC.Address(nil)
	if err != nil {
		log.Fatal(err)
	}

	elClient, err := lwk.ElectrumClientFromUrl(node.ElectrumUrl())
	if err != nil {
		panic(err)
	}

	txid := node.SendToAddress(addr0.Address(), 10000, nil)
	_, err = wolletC.WaitForTx(txid, elClient)
	if err != nil {
		log.Fatal(err)
	}

	// ANCHOR: multisig-send
	// Carol creates a transaction send few sats to a certain address
	address, err := lwk.NewAddress("<address>")
	address = node.GetNewAddress() // ANCHOR: ignore
	err = nil                      // ANCHOR: ignore
	if err != nil {
		log.Fatal(err)
	}
	sendSats := uint64(1000)
	lbtc := network.PolicyAsset()

	builder := network.TxBuilder()
	if err := builder.AddRecipient(address, sendSats, lbtc); err != nil {
		log.Fatal(err)
	}
	pset, err := builder.Finish(wolletC)
	if err != nil {
		log.Fatal(err)
	}

	pset, err = signerC.Sign(pset)
	if err != nil {
		log.Fatal(err)
	}

	// Carol sends the PSET to Bob
	// Bob wants to analyze the PSET before signing, thus he creates a wollet
	wdB, err := lwk.NewWolletDescriptor(descStr)
	if err != nil {
		log.Fatal(err)
	}
	wolletB, err := lwk.NewWollet(network, wdB, nil)
	if err != nil {
		log.Fatal(err)
	}

	updateB, err := client.FullScan(wolletB)
	if err != nil {
		log.Fatal(err)
	}
	if updateB != nil {
		if err := wolletB.ApplyUpdate(*updateB); err != nil {
			log.Fatal(err)
		}
	}

	// Then Bob uses the wollet to analyze the PSET
	details, err := wolletB.PsetDetails(pset)
	if err != nil {
		log.Fatal(err)
	}

	// PSET has a reasonable fee
	if details.Balance().FeesIn(lbtc) >= 100 {
		panic("fee too high")
	}

	// PSET has a signature from Carol
	fingerprintsHas := details.FingerprintsHas()
	if len(fingerprintsHas) != 1 {
		panic("unexpected fingerprints_has len")
	}

	fpC, err := signerC.Fingerprint()
	if err != nil {
		log.Fatal(err)
	}
	hasC := false
	for _, fp := range fingerprintsHas {
		if fp == fpC {
			hasC = true
			break
		}
	}
	if !hasC {
		panic("missing signer_c fingerprint")
	}

	// PSET needs a signature from either Bob or Carol
	fingerprintsMissing := details.FingerprintsMissing()
	if len(fingerprintsMissing) != 2 {
		panic("unexpected fingerprints_missing len")
	}
	fpA, err := signerA.Fingerprint()
	if err != nil {
		log.Fatal(err)
	}
	fpB, err := signerB.Fingerprint()
	if err != nil {
		log.Fatal(err)
	}

	hasMissingA := false
	hasMissingB := false
	for _, fp := range fingerprintsMissing {
		if fp == fpA {
			hasMissingA = true
		}
		if fp == fpB {
			hasMissingB = true
		}
	}
	if !hasMissingA || !hasMissingB {
		panic("missing signer_a or signer_b fingerprint")
	}

	// PSET has a single recipient, with data matching what was specified above
	recipients := details.Balance().Recipients()
	if len(recipients) != 1 {
		panic("unexpected recipients len")
	}
	recipient := recipients[0]
	if fmt.Sprintf("%v", *recipient.Address()) != fmt.Sprintf("%v", address) {
		panic("unexpected recipient address")
	}
	if *recipient.Asset() != lbtc {
		panic("unexpected recipient asset")
	}
	if *recipient.Value() != sendSats {
		panic("unexpected recipient value")
	}

	// Bob is satisified with the PSET and signs it
	pset, err = signerB.Sign(pset)
	if err != nil {
		log.Fatal(err)
	}

	// Bob sends the PSET back to Carol
	// Carol checks that the PSET has enough signatures
	detailsC, err := wolletC.PsetDetails(pset)
	if err != nil {
		log.Fatal(err)
	}
	if len(detailsC.FingerprintsHas()) != 2 {
		panic("unexpected fingerprints_has len")
	}

	// Carol finalizes the PSET and broadcast the transaction
	tx, err := pset.Finalize()
	if err != nil {
		log.Fatal(err)
	}
	txid, err = client.Broadcast(tx)
	if err != nil {
		log.Fatal(err)
	}
	// ANCHOR_END: multisig-send

	fmt.Println("Broadcast multisig txid:", txid)
}
